[dtensor] add a implicit replication flag

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115566
* #115565
* #115564
* __->__ #115297

This PR adds a experimental implicit replication support for DTensor to
inter-op with torch.Tensor, basically under this context manager DTensor
could work together with torch.Tensor by assuming the torch.Tensor
sharding layout is replicated.

Note that this is risky for DTensor so we don't turn it on by default,
but for certain cases where it is for sure replicated, user can use this
to allow DTensor and Tensor computation work together

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @lucasllc

wanchaol committed:
[dtensor] add a implicit replication flag

This PR adds a experimental implicit replication support for DTensor to
inter-op with torch.Tensor, basically under this context manager DTensor
could work together with torch.Tensor by assuming the torch.Tensor
sharding layout is replicated.

Note that this is risky for DTensor so we don't turn it on by default,
but for certain cases where it is for sure replicated, user can use this
to allow DTensor and Tensor computation work together

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add a implicit replication flag"


This PR adds a experimental implicit replication support for DTensor to
inter-op with torch.Tensor, basically under this context manager DTensor
could work together with torch.Tensor by assuming the torch.Tensor
sharding layout is replicated.

Note that this is risky for DTensor so we don't turn it on by default,
but for certain cases where it is for sure replicated, user can use this
to allow DTensor and Tensor computation work together

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 kiukchung lucasllc

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add a implicit replication flag"


This PR adds a experimental implicit replication support for DTensor to
inter-op with torch.Tensor, basically under this context manager DTensor
could work together with torch.Tensor by assuming the torch.Tensor
sharding layout is replicated.

Note that this is risky for DTensor so we don't turn it on by default,
but for certain cases where it is for sure replicated, user can use this
to allow DTensor and Tensor computation work together

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 kiukchung lucasllc

[ghstack-poisoned]

awgu reviewed (approved):
Thanks for adding this!

awgu commented on test/distributed/_tensor/test_dtensor.py:
```diff
@@ -742,6 +742,18 @@ def test_redistribute_sub_mesh(self):
+
+        with implicit_replication():
+            out_dt = sharded_dtensor + torch.randn(3, device=self.device_type)
+            self.assertEqual(out_dt.placements, [Shard(0)])
```
Should we include any numeric test to check the result of the add?

What is the expected semantics of adding a replicated `(3,)` local tensor to a `(4, 3)` local tensor? It should broadcast over dim-1?

wanchaol replied:
sure!

> What is the expected semantics of adding a replicated (3,) local tensor to a (4, 3) local tensor? It should broadcast over dim-1?

Yep it would be implicitly broadcasted according on the "4" dim according to the tensor broadcasting rule https://pytorch.org/docs/stable/notes/broadcasting.html#broadcasting-semantics

wanchaol committed:
Update on "[dtensor] add a implicit replication flag"


This PR adds a experimental implicit replication support for DTensor to
inter-op with torch.Tensor, basically under this context manager DTensor
could work together with torch.Tensor by assuming the torch.Tensor
sharding layout is replicated.

Note that this is risky for DTensor so we don't turn it on by default,
but for certain cases where it is for sure replicated, user can use this
to allow DTensor and Tensor computation work together

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 kiukchung lucasllc

[ghstack-poisoned]
